### PR TITLE
feat: Required expression so required can be conditional

### PR DESCRIPTION
**Pull request by harmbrugge**

### CODE REVIEW

File: apps/schema/src/components/ColumnEditModal.vue [hunk 1]
```diff
@@ -114,11 +114,18 @@
...
             <div class="row">
               <div class="col-4" v-if="isEditable(column)">
-                <InputBoolean
-                  id="column_required"
-                  v-model="column.required"
+                <InputRadio
+                  id="column_required_radio"
                   label="required"
+                  :options="[true, false, 'condition']"
+                  v-model="requiredSelect"
```
Comment (mswertz) [inline apps/schema/src/components/ColumnEditModal.vue:121]: ik kijk vast niet goed, maar hoe komt als 'true' or 'false' de waarde in column.required terecht?

File: backend/molgenis-emx2-io/src/test/java/org/molgenis/emx2/io/TestImportEmptyAutoId.java [hunk 2]
```diff
@@ -0,0 +1,57 @@
+package org.molgenis.emx2.io;
```
Comment (mswertz) [inline backend/molgenis-emx2-io/src/test/java/org/molgenis/emx2/io/TestImportEmptyAutoId.java:1]: other PR

**Review by mswertz — COMMENTED** [review]
first review looks good, only  minor comments. 
I wait until 'green' to also test it :-)

File: apps/schema/src/components/ColumnEditModal.vue (hunk 1 above)
Comment (harmbrugge) [reply apps/schema/src/components/ColumnEditModal.vue:121]: @update:modelValue="setRequired" dus in deze functie wordt ie gezet

File: backend/molgenis-emx2-graphql/src/main/java/org/molgenis/emx2/json/Column.java [hunk 3]
```diff
@@ -160,14 +160,22 @@ public void setKey(Integer key) {
     this.key = key;
   }
 
-  public Boolean getRequired() {
-    return required;
+  public boolean isRequired() {
+    return required != null && required.equals("true");
```
[the remark this answers is not in the document]
Comment (harmbrugge) [reply backend/molgenis-emx2-graphql/src/main/java/org/molgenis/emx2/json/Column.java:164]: Als deze override er niet is gaat veel bestaande code stuk

File: backend/molgenis-emx2-sql/src/main/java/org/molgenis/emx2/sql/SqlTypeUtils.java [hunk 4]
```diff
@@ -89,21 +89,25 @@ private static void applyAutoid(Column c, Row row) {
...
+          }
+        } else {
+          if (row.isNull(c.getName(), c.getColumnType())) {
             throw new MolgenisException("column '" + c.getName() + "' is required in " + row);
           }
         }
-      } else {
-        if (row.isNull(c.getName(), c.getColumnType())) {
-          throw new MolgenisException("column '" + c.getName() + "' is required in " + row);
+      } else if (c.isConditionallyRequired()) {
```
Comment (mswertz) [inline backend/molgenis-emx2-sql/src/main/java/org/molgenis/emx2/sql/SqlTypeUtils.java:106]: I found 'isRequired' vs 'isConditionallyRequired' a bit confusing initially but it is growing on me now :-)

**Review by mswertz — COMMENTED** [review]
I can't find a junit test that uses the conditionallyRequired?
That is also why 'sonar' is giving an error

**Review by mswertz — CHANGES_REQUESTED** [review]
I couldn't find the test for the conditionalRequired == true (so basically, a test with a validation expression)

And we need a bit of docs here on this new feature https://github.com/molgenis/molgenis-emx2/blob/master/docs/molgenis/use_schema.md#required

File: backend/molgenis-emx2-sql/src/main/java/org/molgenis/emx2/sql/SqlTypeUtils.java [hunk 5]
```diff
@@ -89,21 +89,25 @@ private static void applyAutoid(Column c, Row row) {
...
-        && c.getComputed() == null
-        && !AUTO_ID.equals(c.getColumnType())
-        && c.isRequired()) {
-
-      if (c.isReference()) {
-        for (Reference r : c.getReferences()) {
-          if (row.isNull(r.getName(), r.getPrimitiveType())) {
+  private static void checkRequired(Column c, Row row, Map<String, Object> values) {
+    if (!row.isDraft() && c.getComputed() == null && !AUTO_ID.equals(c.getColumnType())) {
+      if (c.isRequired()) {
```
Comment (mswertz) [inline backend/molgenis-emx2-sql/src/main/java/org/molgenis/emx2/sql/SqlTypeUtils.java:94]: I think this is still incorrect I believe. I think it should be one 'if' without your new else that looks something like

`if(c.isRequired || executeRequiredExpression(c.getExpression)) { 
... //here the original check code to see if the values are not null 
}`

with something like
`boolean executeRequiredExpression(Column c, Map values) {
  String result = checkValidation(c.getRequired(), values);
  return result != null && !"false".equals(result)
}
`

File: backend/molgenis-emx2-sql-it/src/test/java/org/molgenis/emx2/sql/TestConditionalRequired.java [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+import static org.molgenis.emx2.TableMetadata.table;
+
+import org.junit.jupiter.api.Test;
+import org.molgenis.emx2.*;
+
+public class TestConditionalRequired {
+
+  @Test
+  public void testConditionallyRequiredOnSingleField() {
+    String requiredExpression = "if(field_one) 'if field_one is provided field_two is required'";
```
Comment (mswertz) [inline backend/molgenis-emx2-sql-it/src/test/java/org/molgenis/emx2/sql/TestConditionalRequired.java:14]: ah, leuk idee om mogelijk te maken een message te vuren. Want tot nu toe benaderde ik het als een manier om de standaard required message terug te geven, maar het heeft inderdaad wel meerwaarde om iets moois te kunnen zeggen..

File: backend/molgenis-emx2-sql-it/src/test/java/org/molgenis/emx2/sql/TestConditionalRequired.java [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+        assertThrows(
+            MolgenisException.class,
+            () -> {
+              SqlTypeUtils.applyValidationAndComputed(tableMetadata.getColumns(), invalidRow);
+            });
+
+    assertEquals(
+        exception.getMessage(),
+        "column 'medicalStatus' is required: "
+            + "Medical status should be provided when an old fat cat is on drugs "
```
Comment (mswertz) [inline backend/molgenis-emx2-sql-it/src/test/java/org/molgenis/emx2/sql/TestConditionalRequired.java:77]: LOL

**Comment by harmbrugge** [comment]
I think all comments are addressed now; added documentation and tests

**Review by mswertz — COMMENTED** [review]
Nearly there.

* I would like to see test where my script doesn't return an error string but simply a 'true' or a 'false'. In those cases I expect the default required message when 'true' and when 'false' to skip the hasEmptyFields check. 

* see above option also in the manual

*  I still see stuff from the autoid commit here. Maybe a `git pull origin master` is all you need?

**Review by mswertz — CHANGES_REQUESTED** [review]
code looked good.

However, form didn't show the message (while the source code seems to have logic for it so I don't understand)

When playing with schema editor it didn't respond to metadata changes, e.g. 
image

Also I didn't see this setting in schema view. I expected a 'required = ....' in this screen for the status I just changed
image

And I can even save it 
image

Can it be that empty strings are not evaluated agains missing?
If it works on your machine we should sit together

I tested against 
https://preview-emx2-pr-3152.dev.molgenis.org/pet%20store/schema/#/

**Review by mswertz — APPROVED** [review]

**Pull request merged**